File: src/main/java/innexgo/hours/SessionDataService.java
/*
 * Innexgo Website
 * Copyright (C) 2020 Innexgo LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package innexgo.hours;

import java.util.stream.Stream;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@Repository
public class SessionDataService {

  @Autowired private JdbcTemplate jdbcTemplate;

  public SessionData getBySessionDataId(long sessionDataId) {
    String sql =
        "SELECT * FROM session_data WHERE session_data_id=?";
    RowMapper<SessionData> rowMapper = new SessionDataRowMapper();
    SessionData sessionData = jdbcTemplate.queryForObject(sql, rowMapper, sessionDataId);
    return sessionData;
  }

  public long nextId() {
    String sql = "SELECT max(sesd.session_data_id) FROM session_data sesd";
    Long maxId = jdbcTemplate.queryForObject(sql, Long.class);
    if(maxId == null) {
      return 0;
    } else {
      return maxId + 1;
    }
  }



  public void add(SessionData sessionData) {
    sessionData.sessionDataId = nextId();
    // Add sessionData
    String sql =
        "INSERT INTO session_data values (?, ?, ?, ?, ?, ?, ?, ?, ?)";
    jdbcTemplate.update(
        sql,
        sessionData.sessionDataId,
        sessionData.creationTime,
        sessionData.creatorUserId,
        sessionData.sessionId,
        sessionData.name,
        sessionData.startTime,
        sessionData.duration,
        sessionData.hidden,
        sessionData.active
    );
  }

  public boolean existsBySessionDataId(long sessionDataId) {
    String sql = "SELECT count(*) FROM session_data sesd WHERE sesd.session_data_id=?";
    int count = jdbcTemplate.queryForObject(sql, Integer.class, sessionDataId);
    return count != 0;
  }

  // Restrict sessionDatas by
  public Stream<SessionData> query(
      Long sessionDataId,
      Long creationTime,
      Long minCreationTime,
      Long maxCreationTime,
      Long creatorUserId,
      Long sessionId,
      String name,
      String partialName,
      Long startTime,
      Long minStartTime,
      Long maxStartTime,
      Long duration,
      Long minDuration,
      Long maxDuration,
      Boolean hidden,
      Boolean active,
      boolean onlyRecent,
      long offset,
      long count) {
    String sql =
        "SELECT sesd.* FROM session_data sesd"
            + (!onlyRecent ? "" : " INNER JOIN (SELECT max(session_data_id) id FROM session_data GROUP BY session_id) maxids ON maxids.id = sesd.session_data_id")
            + " WHERE 1=1 "
            + (sessionDataId   == null ? "" : " AND sesd.session_data_id = " + sessionDataId)
            + (creationTime    == null ? "" : " AND sesd.creation_time = " + creationTime)
            + (minCreationTime == null ? "" : " AND sesd.creation_time > " + minCreationTime)
            + (maxCreationTime == null ? "" : " AND sesd.creation_time < " + maxCreationTime)
            + (creatorUserId   == null ? "" : " AND sesd.creator_id = " + creatorUserId)
            + (sessionId       == null ? "" : " AND sesd.session_id = " + sessionId)
            + (name            == null ? "" : " AND sesd.name = " + Utils.escape(name))
            + (partialName     == null ? "" : " AND sesd.name LIKE " + Utils.escape("%"+partialName+"%"))
            + (startTime       == null ? "" : " AND sesd.start_time = " + startTime)
            + (minStartTime    == null ? "" : " AND sesd.start_time > " + minStartTime)
            + (maxStartTime    == null ? "" : " AND sesd.start_time < " + maxStartTime)
            + (duration        == null ? "" : " AND sesd.duration = " + duration)
            + (minDuration     == null ? "" : " AND sesd.duration > " + minDuration)
            + (maxDuration     == null ? "" : " AND sesd.duration < " + maxDuration)
            + (hidden          == null ? "" : " AND sesd.hidden = " + hidden)
            + (active          == null ? "" : " AND sesd.active = " + active)
            + (" ORDER BY sesd.session_data_id")
            + (" LIMIT " + offset + ", " + count)
            + ";";

    RowMapper<SessionData> rowMapper = new SessionDataRowMapper();
    return this.jdbcTemplate.query(sql, rowMapper).stream();
  }
}

File: src/session_data_service.rs
use super::db_types::*;
use super::utils::current_time_millis;
use std::convert::From;
use tokio_postgres::GenericClient;

impl From<tokio_postgres::row::Row> for SessionData {
  // select * from session_data order only, otherwise it will fail
  fn from(row: tokio_postgres::Row) -> SessionData {
    SessionData {
      session_data_id: row.get("session_data_id"),
      creation_time: row.get("creation_time"),
      creator_user_id: row.get("creator_user_id"),
      session_id: row.get("session_id"),
      name: row.get("name"),
      start_time: row.get("start_time"),
      end_time: row.get("end_time"),
      active: row.get("active"),
    }
  }
}

// TODO we nsd to figure out a way to make scheduled and unscheduled goals work better
pub async fn add(
  con: &mut impl GenericClient,
  creator_user_id: i64,
  session_id: i64,
  name: String,
  start_time: i64,
  end_time: i64,
  active: bool,
) -> Result<SessionData, tokio_postgres::Error> {
  let creation_time = current_time_millis();

  let session_data_id = con
    .query_one(
      "INSERT INTO
       session_data(
           creation_time,
           creator_user_id,
           session_id,
           name,
           start_time,
           end_time,
           active
       )
       VALUES ($1, $2, $3, $4, $5, $6, $7)
       RETURNING session_data_id
      ",
      &[
        &creation_time,
        &creator_user_id,
        &session_id,
        &name,
        &start_time,
        &end_time,
        &active,
      ],
    )
    .await?
    .get(0);

  Ok(SessionData {
    session_data_id,
    creation_time,
    creator_user_id,
    session_id,
    name,
    start_time,
    end_time,
    active,
  })
}

#[allow(unused)]
pub async fn get_by_session_data_id(
  con: &mut impl GenericClient,
  session_data_id: &i64,
) -> Result<Option<SessionData>, tokio_postgres::Error> {
  let result = con
    .query_opt(
      "SELECT * FROM session_data WHERE session_data_id=$1",
      &[&session_data_id],
    )
    .await?
    .map(|x| x.into());

  Ok(result)
}

pub async fn query(
  con: &mut impl GenericClient,
  props: innexgo_hours_api::request::SessionDataViewProps,
) -> Result<Vec<SessionData>, tokio_postgres::Error> {

  let sql = [
    "SELECT sesd.* FROM session_data sesd",
    if props.only_recent {
      " INNER JOIN
          (SELECT max(session_data_id) id FROM session_data GROUP BY session_id) maxids
          ON maxids.id = sesd.session_data_id"
    } else {
      ""
    },
    " INNER JOIN session ses ON sesd.session_id = ses.session_id",
    " WHERE 1 = 1",
    " AND ($1::bigint[]  IS NULL OR sesd.session_data_id IN $1)",
    " AND ($2::bigint    IS NULL OR sesd.creation_time >= $2)",
    " AND ($3::bigint    IS NULL OR sesd.creation_time <= $3)",
    " AND ($4::bigint[]  IS NULL OR sesd.creator_user_id IN $4)",
    " AND ($5::bigint[]  IS NULL OR sesd.session_id IN $5)",
    " AND ($6::text[]    IS NULL OR sesd.name IN $6)",
    " AND ($7::text      IS NULL OR sesd.name LIKE CONCAT('%',$7,'%'))",
    " AND ($8::bigint    IS NULL OR sesd.start_time >= $8)",
    " AND ($9::bigint    IS NULL OR sesd.start_time <= $9)",
    " AND ($10::bigint   IS NULL OR sesd.end_time >= $10)",
    " AND ($11::bigint   IS NULL OR sesd.end_time <= $11)",
    " AND ($12::bool     IS NULL OR sesd.active = $12)",
    " AND ($13::bigint[] IS NULL OR ses.course_id IN $13)",
    " ORDER BY sesd.session_data_id",
  ]
  .join("");

  let stmnt = con.prepare(&sql).await?;

  let results = con
    .query(
      &stmnt,
      &[
        &props.session_data_id,
        &props.min_creation_time,
        &props.max_creation_time,
        &props.creator_user_id,
        &props.session_id,
        &props.name,
        &props.partial_name,
        &props.min_start_time,
        &props.max_start_time,
        &props.min_end_time,
        &props.max_end_time,
        &props.active,
        &props.course_id,
      ],
    )
    .await?
    .into_iter()
    .map(|row| row.into())
    .collect();

  Ok(results)
}
